### DISCUSSION TITLE
Print custom A4 layouts using Dynamic DB data

### DISCUSSION THREAD

**Discussion by Yamita** [directus/directus#18556] [Feature Requests]
### Summary

Hi Team,

I have searched the internet for a tool that can do something like Adobe InDesign with dynamic data directly from your own DB.

### Motivation

Picture this, Directus has a new tab called [Documents] where you can create A4 templates using UI elements which can dynamically update based on data in your DB. Then you can save your layouts in the [Documents] tab and flick through the one you need: invoice, packing slip, return form, whatever. This would allow you to create beautiful invoice templates and print them as PDF or word docs directly from Directus. You would destroy the competition with this feature.

### Detailed Design

I suggest starting with A4 layout and adding more layers in the future. Being able to add header/footer as well so anything placed in the footer will print on the FINAL PAGE if your table in the middle was dynamic and grabbed 50 SKU rows you might print 4 pages but your TnC would be placed in the footer so it always prints that at the end and you get a clean look. I can't code but would be happy to contribute for free with the business aspect and design.

### Must Have:
1) A blank canvas like AdobeXD, Figma or PowerBI
2) header/footer that give UI elements inheritance behavior like any UI element placed in footer appears at the bottom. 
3) UI elements that can be resized and store dynamic data.

### Should Have:
4) You want a stand alone 'PK selector' above the canvas so you can see how your printable document changes e.g. the PK could be a INV# as the key and the value is INV# 300 to test how it dynamically fetches the data from the DB for this INV# and renders it on the canvas layout your created.
5) has to be able to grab relational data INV# and SKU# are most likely stored in different normalized tables. So if you pick INV# it has to grab the correct 20 SKUs in your table element.

### Could Have:
6) Just wanted to add, you don't need all the tooling of Adobe InDesign rather you can make parent elements like header/footer a type of Div that gives inheritance to other UI elements which can be edited with CSS by the user. So you click the UI element and get a small CSS menu. This way not only your data but also your layout becomes dynamic for the printable document.

### Won't Have:
Canvas UI page shouldn't have a ton of UI configuration buttons (at least not for MVP), Instead, pick from a series of default blank UI element you drag onto the canvas like squares, tables, titles, signature field, On the right hand side you have 2 fields when you click a UI element [query] and [CSS] where you can choose what data it fetches and how it looks including font in the CSS (maybe a global font drop down that forces all UI elements to inherit that font style/size). Everything will be 2D object don't need it to be 3D moveable CSS since your intention is to print it. I think that simplifies the project scope a lot.

### Drawbacks

Not sure how you would fetch the data you want to show in each UI element, maybe with a SQL query of maybe with JS code. I think SQL is better if possible.

### Alternatives

I don't know a single one, if you do, I'd love to know.

### Adoption Strategy

It would simply be a [documents] 'feature' of Directus.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
